### PR TITLE
refactor: define node matchers on assertion classes

### PR DESCRIPTION
**Pull request by G-Rath**
I figured this would be enough but didn't want to overload #1790.

It does send the fixer for `InternalAffairs/NodeMatcherDirective` into an inf. loop though - it looks like one part of the cop sees just the "self" whereas the other sees the whole definition, so it reports "use self.match_assertion instead of self" and then the fixer just keeps adding `.match_assertion` to the end of the string until it errors; for now I've just disabled it, but I might try and look into it at some point.

______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [x] ~Updated documentation.~
- [x] ~Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.~
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have created a new cop:

- [x] ~Added the new cop to `config/default.yml`.~
- [x] ~The cop is configured as `Enabled: pending` in `config/default.yml`.~
- [x] ~The cop is configured as `Enabled: true` in `.rubocop.yml`.~
- [x] ~The cop documents examples of good and bad code.~
- [x] ~The tests assert both that bad code is reported and that good code is not reported.~
- [x] ~Set `VersionAdded: "<<next>>"` in `default/config.yml`.~

If you have modified an existing cop's configuration options:

- [x] ~Set `VersionChanged: "<<next>>"` in `config/default.yml`.~

### CODE REVIEW

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 1]
```diff
@@ -211,17 +224,9 @@ def assertion
...
-
-            def_node_matcher "minitest_#{name}".to_sym, m::NODE_MATCHER_PATTERN
-          end
-
           def on_send(node)
             ASSERTION_MATCHERS.each do |m|
-              name = m.name.split('::').last
-
-              public_send("minitest_#{name}".to_sym, node) do |*args|
+              m.match_assertion(node) do |*args|
```
Comment (G-Rath) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:229]: at this point the names are getting a little silly - I'll next explore if we can nicely have the classes handle calling their `match` logic and just return a result

**Comment by bquorning** [comment]
@G-Rath I pushed an alternative implementation at the same time as you opened your PR: #1794. I hadn’t considered using class methods for the node matchers, that’s a neat idea.

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 2]
```diff
@@ -60,7 +66,8 @@ class EqualAssertion < BasicAssertion
               refute_equal
             ].freeze
 
-            NODE_MATCHER_PATTERN = <<~PATTERN
+            # @!method self.minitest_assertion(node)
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:69]: I’m actually not sure if the correct YARD is your suggestion or
```suggestion
            # @!method minitest_assertion(node)
            # @!scope class
```

**Review by bquorning — APPROVED** [review]

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb (hunk 2 above)
Comment (G-Rath) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:69]: I think you're right, but it still annoys `InternalAffairs/NodeMatcherDirective` 😞

Comment (bquorning) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:69]: I opened an issue on RuboCop: https://github.com/rubocop/rubocop/issues/12665

For now, should we consider your PR ready to merge?

Comment (G-Rath) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:69]: yup works for me

**Pull request merged**